### PR TITLE
API/TST: expand tests for string any/all reduction + fix pyarrow-based implementation

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
While working on adding any/all support for the object-dtype version of the StringDtype in https://github.com/pandas-dev/pandas/pull/58451, I bumped into some issues with the current testing and implementation.

It seemed that our current testing is a bit limited, so I expanded the existing test to cover more cases and to be parametrized over all string dtype variants (including plain object dtype).

But that uncovered some issues with the current implementation of the pyarrow-backed version:

- `any` could return `pd.NA` in case of `skipna=False`, while for this version of the string dtype (not using Kleene logic), the result should always be True or False. This is an easy fix (we currently were only filling missing values in case of `all`, also do this for `any`)
- **How to treat missing values?** As truthy or falsey? The current pyarrow-based implementation explicitly filled the missing values with False, but the [documentation](https://pandas.pydata.org/docs/reference/api/pandas.Series.any.html) says _"If skipna is False, then NA are treated as True, because these are not equal to zero."_ 
  I find that a bit strange (my first expectation would be to treat missing values as False), but it _is_ explicitly documented, and follows the behaviour of current object dtype (eg  `pd.Series(["a". np.nan], dtype=object).all(skipna=False)` gives True, not False.

So for now this PR updated the pyarrow implementation to follow the documented behaviour.

- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/arrays/string_arrow.py [hunk 1]
```diff
@@ -697,9 +697,9 @@ def _reduce(
         self, name: str, *, skipna: bool = True, keepdims: bool = False, **kwargs
     ):
         if name in ["any", "all"]:
-            if not skipna and name == "all":
-                nas = pc.invert(pc.is_null(self._pa_array))
-                arr = pc.and_kleene(nas, pc.not_equal(self._pa_array, ""))
+            if not skipna:
+                nas = pc.is_null(self._pa_array)
+                arr = pc.or_kleene(nas, pc.not_equal(self._pa_array, ""))
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/string_arrow.py:702]: Before, for converting the string array to a boolean array, True values were considered as _"not `""`, and not null"_, and this diff changed that to _"not `""`, or null"_

File: pandas/tests/reductions/test_reductions.py [hunk 2]
```diff
@@ -1062,25 +1062,62 @@ def test_any_all_datetimelike(self):
...
+        assert ser.any(skipna=False)
+        assert not ser.all(skipna=False)
 
-        ser = Series([None, "a"], dtype="string[pyarrow_numpy]")
+        ser = Series([np.nan, "a"], dtype=any_string_dtype)
         assert ser.any()
         assert ser.all()
-        assert not ser.all(skipna=False)
+        assert ser.any(skipna=False)
+        assert ser.all(skipna=False)  # NaN is considered truthy
```
Comment (jorisvandenbossche) [inline pandas/tests/reductions/test_reductions.py:1090]: The places where I currently commented this `# NaN is considered truthy` are the test cases where the result would change from True to False if NaN would be considered as falsey instead.

**Comment by mroeschke** [comment]
In 3.0, adjacently `.idxmax/min` will raise a `ValueError` if `skipna=False` and a NA value is encountered. There _could_ be an API behavior argument to align with that behavior in the future.

Besides that point, I would also agree with your initial expectation that NA values are falsey.

**Comment by jorisvandenbossche** [comment]
> In 3.0, adjacently `.idxmax/min` will raise a `ValueError` if `skipna=False` and a NA value is encountered. There _could_ be an API behavior argument to align with that behavior in the future.

For `any`/`all`, I think in the future we don't have to raise about missing values, but we can use Kleene logic and return pd.NA in some cases (although that might then just give an error downstream in your code ..).

That's the behaviour that is currently implemented for the nullable BooleanDtype, and if we plan to stick to that behaviour, I would say it's not worth to change the current NaN-based dtypes to start raise an error for NaNs (and we can preserve the strange "NaN is considered as True because it is not equal to 0" for now).

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jorisvandenbossche

**Pull request merged**